File: csharp-calculator/MainWindow.cs
using Gtk;
using System.Linq;
using System.Collections.Generic;
using UI = Gtk.Builder.ObjectAttribute;
using System;

namespace csharp
{
    class MainWindow : Window
    {

        [UI] private Button numpad0 = null;
        [UI] private Button numpad1 = null;
        [UI] private Button numpad2 = null;
        [UI] private Button numpad3 = null;
        [UI] private Button numpad4 = null;
        [UI] private Button numpad5 = null;
        [UI] private Button numpad6 = null;
        [UI] private Button numpad7 = null;
        [UI] private Button numpad8 = null;
        [UI] private Button numpad9 = null;
        [UI] private Button numpad_add = null;
        [UI] private Button numpad_sub = null;
        [UI] private Button numpad_mul = null;
        [UI] private Button numpad_div = null;
        [UI] private Button numpad_dot = null;
        [UI] private Button numpad_eq = null;
        [UI] private Button numpad_negate = null;
        [UI] private Button numpad_sqrt = null;
        [UI] private Button numpad_square = null;
        [UI] private Button numpad_cube = null;
        [UI] private Button numpad_reciprocal = null;
        [UI] private Button numpad_percent = null;
        [UI] private Button ClearGlobal = null;
        [UI] private Button ClearEntry = null;
        [UI] private Button Backspace = null;
        [UI] private Label expression_label = null;
        [UI] private Entry input = null;

        public MainWindow() : this(new Builder("Main.glade")) { }


        private enum Operation
        {
            ADD, SUB, MUL, DIV, NONE
        }

        private string stringFromOperation(Operation opp)
        {
            switch (opp)
            {
                case Operation.ADD: return " + ";
                case Operation.SUB: return " - ";
                case Operation.MUL: return " × ";
                case Operation.DIV: return " ÷ ";
                default: return "";
            }
        }

        private class Expr
        {
            public double value;
            public Operation operation;
        }

        List<Expr> expression = new List<Expr>();
        bool preview = false;

        private MainWindow(Builder builder) : base(builder.GetRawOwnedObject("MainWindow"))
        {
            DeleteEvent += Window_DeleteEvent;
            builder.Autoconnect(this);

            this.KeyPressEvent += (e, a) =>
            {
                switch (a.Event.Key)
                {
                    case Gdk.Key.Key_0: input.GrabFocus(); break;
                    case Gdk.Key.Key_1: input.GrabFocus(); break;
                    case Gdk.Key.Key_2: input.GrabFocus(); break;
                    case Gdk.Key.Key_3: input.GrabFocus(); break;
                    case Gdk.Key.Key_4: input.GrabFocus(); break;
                    case Gdk.Key.Key_5: input.GrabFocus(); break;
                    case Gdk.Key.Key_6: input.GrabFocus(); break;
                    case Gdk.Key.Key_7: input.GrabFocus(); break;
                    case Gdk.Key.Key_8: input.GrabFocus(); break;
                    case Gdk.Key.Key_9: input.GrabFocus(); break;
                    case Gdk.Key.plus: numpad_add.Activate(); break;
                    case Gdk.Key.minus: numpad_sub.Activate(); break;
                    case Gdk.Key.asterisk: numpad_mul.Activate(); break;
                    case Gdk.Key.slash: numpad_div.Activate(); break;
                    case Gdk.Key.BackSpace: Backspace.Activate(); break;
                    case Gdk.Key.KP_Enter: numpad_eq.Activate(); break;
                    case Gdk.Key.equal: numpad_eq.Activate(); break;
                }
            };

            input.Changed += (object o, EventArgs e) =>
            {
                if (expression_label.Text == "Invalid Number")
                {
                    expression_label.Text = expression
                      .Select(e => e.value.ToString() + stringFromOperation(e.operation))
                      .Aggregate("", (x, accum) => x + accum);
                }
                else if (input.Text.Length > 0)
                {
                    switch (input.Text.Last())
                    {
                        case '+': numpad_add.Activate(); break;
                        case '-': numpad_sub.Activate(); break;
                        case '/': numpad_div.Activate(); break;
                        case '*': numpad_mul.Activate(); break;
                        case '=': numpad_eq.Activate(); break;
                        default: 
                          double val = 0;
                          if (!double.TryParse(input.Text, out val))
                            expression_label.Text = "Invalid Number";
                          return;
                    }
                    input.Text = input.Text.Remove(input.Text.Length - 1);
                }

            };

            numpad0.Clicked += input_digit('0');
            numpad1.Clicked += input_digit('1');
            numpad2.Clicked += input_digit('2');
            numpad3.Clicked += input_digit('3');
            numpad4.Clicked += input_digit('4');
            numpad5.Clicked += input_digit('5');
            numpad6.Clicked += input_digit('6');
            numpad7.Clicked += input_digit('7');
            numpad8.Clicked += input_digit('8');
            numpad9.Clicked += input_digit('9');
            numpad_negate.Clicked += delegate
            {
                if (input.Text.Length == 0) return;
                if (input.Text[0] == '-')
                    input.Text = input.Text.Remove(0, 1);
                else
                    input.Text = '-' + input.Text;

            };
            numpad_dot.Clicked += delegate
            {
                if (!input.Text.Any(x => x == '.'))
                    input.Text += ".";
            };
            numpad_add.Clicked += update_expression(Operation.ADD);
            numpad_sub.Clicked += update_expression(Operation.SUB);
            numpad_mul.Clicked += update_expression(Operation.MUL);
            numpad_div.Clicked += update_expression(Operation.DIV);
            numpad_eq.Clicked += delegate
            {
                if (expression.Count != 0)
                {
                    double val = 0;
                    if (!double.TryParse(input.Text, out val))
                    {
                        if (expression.Last().operation == Operation.MUL
                            || expression.Last().operation == Operation.DIV)
                            val = 1;
                    }
                    expression.Add(new Expr { value = val, operation = Operation.NONE });

                    expression_label.Text = expression
                      .Select(e => e.value.ToString() + stringFromOperation(e.operation))
                      .Aggregate("", (x, accum) => x + accum);
                    input.Text = evaluate_expression().ToString();
                    preview = true;
                    expression = new List<Expr>();
                }
            };

            ClearEntry.Clicked += delegate { input.Text = ""; };
            ClearGlobal.Clicked += delegate
            {
                input.Text = "";
                expression_label.Text = "";
                expression = new List<Expr>();
            };

            Backspace.Clicked += delegate
            {
                if (!preview && input.Text.Length > 0)
                    input.Text = input.Text.Remove(input.Text.Length - 1);
            };

            numpad_sqrt.Clicked += delegate
            {
                double val = 0;
                if (double.TryParse(input.Text, out val))
                {
                    input.Text = (Math.Sqrt(val)).ToString();
                }
            };
            numpad_square.Clicked += delegate
            {
                double val = 0;
                if (double.TryParse(input.Text, out val))
                {
                    input.Text = (Math.Pow(val, 2)).ToString();
                }
            };
            numpad_cube.Clicked += delegate
            {
                double val = 0;
                if (double.TryParse(input.Text, out val))
                {
                    input.Text = (Math.Pow(val, 3)).ToString();
                }
            };
            numpad_reciprocal.Clicked += delegate
            {
                double val = 0;
                if (double.TryParse(input.Text, out val))
                {
                    input.Text = (1 / val).ToString();
                }
            };
            numpad_percent.Clicked += delegate
            {
                double val = 0;
                if (double.TryParse(input.Text, out val))
                {
                    double total = evaluate_expression();
                    input.Text = (total * val / 100).ToString();
                }
            };
        }

        private double evaluate_expression()
        {
            return expression.Aggregate((accum, x) => accum.operation switch
            {
                Operation.ADD => new Expr { value = accum.value + x.value, operation = x.operation },
                Operation.SUB => new Expr { value = accum.value - x.value, operation = x.operation },
                Operation.MUL => new Expr { value = accum.value * x.value, operation = x.operation },
                Operation.DIV => new Expr { value = accum.value / x.value, operation = x.operation },
                _ => throw new NotImplementedException()
            }).value;
        }

        private EventHandler input_digit(char digit)
        {
            return delegate
            {
                if (preview)
                {
                    input.Text = digit.ToString();
                    preview = false;
                }
                else
                    input.Text += digit;
            };
        }

        private EventHandler update_expression(Operation opp)
        {
            return delegate
            {
                double val = 0;
                if (double.TryParse(input.Text, out val))
                    expression.Add(new Expr { value = val, operation = opp });

                if (expression.Count >= 2)
                {
                    input.Text = evaluate_expression().ToString();
                    preview = true;
                }
                else
                    input.Text = "";

                expression_label.Text = expression
                  .Select(e => e.value.ToString() + stringFromOperation(e.operation))
                  .Aggregate("", (x, accum) => x + accum);
            };
        }

        private void Window_DeleteEvent(object sender, DeleteEventArgs a)
        {
            Application.Quit();
        }

    }
}

File: python-calculator/python_calculator/mainWindow.py
import gi
import copy
import math
import functools
from python_calculator.resources import *
from functools import *
from gi.repository import Gtk, Gdk, Gio
from enum import Enum


class Operation(Enum):
    ADD = 0
    SUB = 1
    MUL = 2
    DIV = 3
    NONE = 4


class Expr:
    def __init__(self, val, opp):
        self.val = val
        self.opp = opp

    def __str__(self):
        match self.opp:
            case Operation.ADD:
                return f"{self.val:g} + "
            case Operation.SUB:
                return f"{self.val:g} - "
            case Operation.MUL:
                return f"{self.val:g} × "
            case Operation.DIV:
                return f"{self.val:g} ÷ "
            case Operation.NONE:
                return str(self.val)


class MainWindow(Gtk.Window):
    def __init__(self):
        provider = Gtk.CssProvider()
        provider.load_from_data(style_css)
        Gtk.StyleContext.add_provider_for_screen(
            Gdk.Screen.get_default(), provider, 800
        )

        builder = Gtk.Builder()
        builder.add_from_string(main_glade)

        self.mainWindow = builder.get_object("MainWindow")
        self.mainWindow.connect("destroy", Gtk.main_quit)
        self.numpad0 = builder.get_object("numpad0")
        self.numpad1 = builder.get_object("numpad1")
        self.numpad2 = builder.get_object("numpad2")
        self.numpad3 = builder.get_object("numpad3")
        self.numpad4 = builder.get_object("numpad4")
        self.numpad5 = builder.get_object("numpad5")
        self.numpad6 = builder.get_object("numpad6")
        self.numpad7 = builder.get_object("numpad7")
        self.numpad8 = builder.get_object("numpad8")
        self.numpad9 = builder.get_object("numpad9")
        self.numpad_add = builder.get_object("numpad_add")
        self.numpad_sub = builder.get_object("numpad_sub")
        self.numpad_mul = builder.get_object("numpad_mul")
        self.numpad_div = builder.get_object("numpad_div")
        self.numpad_dot = builder.get_object("numpad_dot")
        self.numpad_eq = builder.get_object("numpad_eq")
        self.numpad_negate = builder.get_object("numpad_negate")
        self.numpad_sqrt = builder.get_object("numpad_sqrt")
        self.numpad_square = builder.get_object("numpad_square")
        self.numpad_cube = builder.get_object("numpad_cube")
        self.numpad_reciprocal = builder.get_object("numpad_reciprocal")
        self.numpad_percent = builder.get_object("numpad_percent")
        self.clearGlobal = builder.get_object("ClearGlobal")
        self.clearEntry = builder.get_object("ClearEntry")
        self.backspace = builder.get_object("Backspace")
        self.expression_label = builder.get_object("expression_label")
        self.input = builder.get_object("input")
        self.expression = []
        self.preview = False

        self.numpad0.connect("clicked", self.input_digit("0"))
        self.numpad1.connect("clicked", self.input_digit("1"))
        self.numpad2.connect("clicked", self.input_digit("2"))
        self.numpad3.connect("clicked", self.input_digit("3"))
        self.numpad4.connect("clicked", self.input_digit("4"))
        self.numpad5.connect("clicked", self.input_digit("5"))
        self.numpad6.connect("clicked", self.input_digit("6"))
        self.numpad7.connect("clicked", self.input_digit("7"))
        self.numpad8.connect("clicked", self.input_digit("8"))
        self.numpad9.connect("clicked", self.input_digit("9"))
        self.clearEntry.connect("clicked", self.clear_entry)
        self.clearGlobal.connect("clicked", self.clear_global)
        self.numpad_add.connect("clicked", self.update_expression(Operation.ADD))
        self.numpad_sub.connect("clicked", self.update_expression(Operation.SUB))
        self.numpad_mul.connect("clicked", self.update_expression(Operation.MUL))
        self.numpad_div.connect("clicked", self.update_expression(Operation.DIV))
        self.numpad_dot.connect("clicked", self.add_dot)
        self.numpad_eq.connect("clicked", self.eq)
        self.backspace.connect("clicked", self.backspace_fn)
        self.numpad_percent.connect("clicked", self.percent)
        self.numpad_negate.connect("clicked", self.negate)
        self.numpad_sqrt.connect("clicked", self.modify_input(math.sqrt))
        self.numpad_square.connect("clicked", self.modify_input(lambda x: x * x))
        self.numpad_cube.connect("clicked", self.modify_input(lambda x: x * x * x))
        self.numpad_reciprocal.connect("clicked", self.modify_input(lambda x: 1/x))

    def percent(self, button):
        try: 
            if len(self.expression) > 0:
                val = float(self.input.get_text())
                self.input.set_text(f'{(self.eval_expression() * (val/100)):g}')
        except ValueError:
            print("invalid input")

    def modify_input(self, fn):
        def f(button):
            try:
                val = float(self.input.get_text())
                self.input.set_text(f'{fn(val):g}')
            except ValueError:
                print("invalid input")
        return f

    def eq(self, button):
        txt = self.input.get_text()
        try: 
            val = float(txt)
            self.expression.append(Expr(val, Operation.NONE))
            self.input.set_text(f'{self.eval_expression():g}')
            self.expression.clear()
            self.expression_label.set_text("")
            self.preview = True
        except ValueError:
            print("invalid input")

    def backspace_fn(self, button):
        txt = self.input.get_text()
        if txt != "":
            self.input.set_text(txt[:-1])
            self.preview = False

    def add_dot(self, button):
        txt = self.input.get_text()
        if not "." in txt:
            self.input.set_text(txt + ".")

    def negate(self, button):
        txt = self.input.get_text()
        if txt.startswith("-"):
            self.input.set_text(txt[1:])
        elif len(txt) > 0:
            self.input.set_text("-" + txt)

    def eval_expression(self):
        def expression_operator(accum, x):
            match accum.opp:
                case Operation.ADD:
                    return Expr(accum.val + x.val, x.opp)
                case Operation.SUB:
                    return Expr(accum.val - x.val, x.opp)
                case Operation.MUL:
                    return Expr(accum.val * x.val, x.opp)
                case Operation.DIV:
                    return Expr(accum.val / x.val, x.opp)
                case Operation.NONE:
                    return Expr(accum.val, x.opp)

        return reduce(expression_operator, self.expression).val

    def update_expression(self, opp):
        def f(button):
            try:
                val = float(self.input.get_text())
                self.expression.append(Expr(val, opp))
                self.expression_label.set_text(
                    functools.reduce(
                        lambda accum, x: accum + str(x), self.expression, ""
                    )
                )

                if len(self.expression) > 1:
                    self.input.set_text(f"{self.eval_expression():g}")
                else:
                    self.input.set_text("")
                self.preview = True
            except:
                None

        return f

    def clear_entry(self, button):
        self.input.set_text("")
        self.preview = False

    def clear_global(self, button):
        self.input.set_text("")
        self.expression.clear()
        self.expression_label.set_text("")
        self.preview = False

    def input_digit(self, chr):
        def f(button):
            txt = self.input.get_text()
            if self.preview:
                self.input.set_text(str(chr))
                self.preview = False
            else:
                self.input.set_text(txt + str(chr))

        return f
